feat: add cli module and traits UserInterface and TodoStorage

AlanDouglasASouza opened the pull request:
Created a cli module that takes responsibility for executing the program through the run method. New traits were also added that abstract the Terminal and Todos methods.

AlanDouglasASouza committed:
feat: add add cli module and traits UserInterface and TodoStorage

Created a cli module that takes responsibility for executing the program through the run method. New traits were also added that abstract the Terminal and Todos methods.

AlanDouglasASouza committed:
refac: the removal of welcome from the main module

The main module is only responsible for calling TodoCli and the presentation layer is responsible for Terminal.

luisholanda commented on src/cli.rs:
```diff
@@ -0,0 +1,88 @@
+        loop {
+            self.user_interface.show_options()?;
+
+            match self.user_interface.get_user_command()? {
```
Esse match tá meio grande, oq você acha de separar cada comando em seu próprio método?

AlanDouglasASouza replied:
Sim, realmente fica mais claro separando por métodos.

luisholanda commented on src/cli.rs:
```diff
@@ -0,0 +1,88 @@
+}
+
+impl TodoCli {
+    pub fn new() -> Self {
+        Self {
+            user_interface: Box::new(Terminal::new()),
+            todo_storage: Box::new(Todos::new()),
+        }
+    }
```
O ideal seria você receber a UI e o storage via parâmetros. Dessa forma a CLI fica genérica sobre os detalhes de como esses são implementados.

AlanDouglasASouza replied:
Ok! Uma duvida: seria correto - ou necessário - ter esse método `new()`? Eu refatorei passando os parâmetros para ele, porém, fiquei com essa duvida. WDYT?

luisholanda replied:
Normalmente criamos um método `new` ou similar, com isso evitamos ter que expor os campos da struct, nos dando mais controle sobre a implementação. Nesse projeto não vai fazer muita diferença ter ou não ter, mas existem situações em que você quer, e.g., ter um cache interno a struct ou outra coisa similar.

AlanDouglasASouza replied:
Show, então esta alinhado com o que eu pensava mesmo. :heavy_check_mark:

luisholanda commented on src/terminal.rs:
```diff
@@ -12,13 +12,40 @@ pub struct Terminal {
+    fn show_error(&self, error: TerminalError);
+    fn get_todo_for_update(
+        &self,
+        list_todos: &Box<dyn TodoStorage>,
```
```suggestion
        list_todos: &dyn TodoStorage,
```

Na maioria das situações, `&Box<_>` é desnecessário.

AlanDouglasASouza replied:
Certo! Para remover o `&Box<_>` eu precisei desreferenciar na chamada ao passar por parâmetro  `check_list_is_empty(&*self.todo_storage)`. Isso está correto?

luisholanda replied:
Acredito que apenas `check_list_is_empty(&self.todo_storage)` já seria suficiente, mas esse tipo de coisa o linter vai indicar quando chegarmos nessa parte 😉

AlanDouglasASouza replied:
Beleza!!

luisholanda commented on src/terminal.rs:
```diff
@@ -108,28 +141,28 @@ impl Terminal {
     }
 
-    pub fn check_list_is_empty(&self, list: &Todos) -> bool {
+    fn check_list_is_empty(&self, list: &Box<dyn TodoStorage>) -> bool {
```
Esse cara fazer parte da UI é meio estranho... Isso não deveria estar na CLI?

AlanDouglasASouza replied:
Concordo! Estava perdido ali... foi alterado. :ok_hand:

luisholanda commented on src/todos.rs:
```diff
@@ -1,36 +1,45 @@
     }
 
-    pub fn show_all_todos(&self, show_keys: bool) -> Result<(), TerminalError> {
+    fn show_all_todos(&self, show_keys: bool) -> Result<(), TerminalError> {
```
Não acho que isso seja responsabilidade do storage... Não seria melhor colocar isso na CLI?

AlanDouglasASouza replied:
Assim o `Todos` nem precisa conhecer o `Terminal`. :ok:

AlanDouglasASouza committed:
refac: adds methods in Todo Cli abstracting from the run method

Abstracts from the match in the run method, in Todo Cli, transferring to methods. Passing storage and user_interface as parameters to make the TodoCli implementation more generic. And removal of Box<_>.

AlanDouglasASouza committed:
refac: removes check_list_is_empty from Terminal

Abstracts the check_list_is_empty method from the Terminal, for TodoCli.

AlanDouglasASouza committed:
refac: move show_all_todos from Todos

Implementation of the get_collection method in the TodoStorage Trait and relocation of the show_all_todos method to TodoCli

luisholanda commented on src/cli.rs:
```diff
@@ -0,0 +1,122 @@
+                UserCommand::Other => self.user_interface.show_invalid_option()?,
+                UserCommand::Exit => {
+                    self.user_interface.finish_todo()?;
+                    break;
```
Nitpicking: dado que esse é o único ponto de saída, não seria mais claro colocar o `return Ok(())` direto aqui?

AlanDouglasASouza replied:
Realmente!

luisholanda commented on src/cli.rs:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    fn check_list_is_empty(&self, list: &dyn TodoStorage) -> bool {
+        if list.len() < 1 {
```
Que tal um `TodoStorage::is_empty`?

```suggestion
        if list.is_empty() {
```

AlanDouglasASouza replied:
Fica mais coerente o nome do método, no começo esse método era pra ser outra coisa e no fim esqueci completamente de alterar o nome dele. Valeu!

luisholanda reviewed (approved):
Só umas observações final, de resto LGTM :D

AlanDouglasASouza committed:
refac: rename method len to is_empty in Todos
